onnxruntime/core/providers/shared_library/provider_author.h

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// Provider authors include this file
+
+#pragma once
+
+#include <vector>
+#include <string>
+#include <map>
+#include <unordered_map>
+#include "onnx/common/stl_backports.h"
+#include "core/common/common.h"
+#include "core/common/const_pointer_container.h"
+#include "core/session/onnxruntime_c_api.h"
+#include "provider_interfaces.h"
+
+namespace ONNX_NAMESPACE {
+
+enum AttributeProto_AttributeType : int {
+  AttributeProto_AttributeType_UNDEFINED = 0,
+  AttributeProto_AttributeType_FLOAT = 1,
+  AttributeProto_AttributeType_INT = 2,
+  AttributeProto_AttributeType_STRING = 3,
+  AttributeProto_AttributeType_TENSOR = 4,
+  AttributeProto_AttributeType_GRAPH = 5,
+  AttributeProto_AttributeType_SPARSE_TENSOR = 11,
+  AttributeProto_AttributeType_FLOATS = 6,
+  AttributeProto_AttributeType_INTS = 7,
+  AttributeProto_AttributeType_STRINGS = 8,
+  AttributeProto_AttributeType_TENSORS = 9,
+  AttributeProto_AttributeType_GRAPHS = 10,
+  AttributeProto_AttributeType_SPARSE_TENSORS = 12
+};
+
+enum OperatorStatus : int {
+  EXPERIMENTAL = 0,
+  STABLE = 1
+};
+
+}  // namespace ONNX_NAMESPACE
+
+namespace onnxruntime {
+
+// The function passed in will be run on provider DLL unload. This is used to free thread_local variables that are in threads we don't own
+// Since these are not destroyed when the DLL unloads we have to do it manually. Search for usage for an example.
+struct RunOnUnload {
+  RunOnUnload(std::function<void()> run);
+  ~RunOnUnload();
+
+ private:
+  bool* enabled_{};
```

Comment:
nit: struct member data don't have trailing underscore.

Reply:
Ah, I just made them structs as they started with public stuff. Should I change them all to classes and add public overrides for everything?